js/runner.go

```diff
@@ -557,9 +557,25 @@ func (r *Runner) runPart(
 		// otherwise we have timeouted
 		return v, newTimeoutError(name, r.getTimeoutFor(name))
 	}
+
 	return v, err
 }
 
+// unPromisify gets the result of v if it is a promise, otherwise returns v
+func unPromisify(v goja.Value) goja.Value {
+	if !isNullish(v) {
+		if p, ok := v.Export().(*goja.Promise); ok {
+			return p.Result()
+		}
+	}
+
+	return v
+}
```

Comment:
Hmm, `goja.Value.Export()` could be a potentially quite costly operation, right? So, ideally, i think that we should either only do it once and then return the result (if it's not a `Promise`), or have some other way of detecting of the `goja.Value` is a `goja.Promise` underneath :thinking: 

Option one could look somewhat like this:

```suggestion
// unPromisifyAndExport gets the result of v if it is a promise, otherwise returns v
func unPromisifyAndExport(v goja.Value) interface{} {
	if isNullish(v) {
		return v
	}
	exp := v.Export()
	if p, ok := exp.(*goja.Promise); !ok {
		return exp
	}
	return p.Result().Export()
}
```

The downside to this is that we'd need to change the `runFn()` signature or call this on every result of `runFn()`, which doesn't seem too bad.

Reply:
While `Export` might be expensive(for some deffinition):
1.  the change you propose needs to ripple through each usafe of `runFn` and `runPart` to matter - which quite a bit bigger change
2. The only 2 functions that usually return are `setup` and `handleSummary`. In the common case of a `default` function not returning anything it will be caught by the `isNullish` check continue. I don't think making `setup` or `handleSummary` a bit more expensive is enough reason to make all the changes `1` entails.

Reply:
[9398f37](https://github.com/grafana/k6/pull/2985/commits/9398f37addf3036c720c0d0489d6e162ed45415b) uses ExportType and reflect to the same check and it seems to work.

This should hopefully also fix any case where `setup` returns too big of an object and `Export` would actually be very expensive.